### DISCUSSION TITLE
Could we import markdown files in a better way?

### DISCUSSION THREAD

**Discussion by AnteeHub** [facebook/docusaurus#8145] [Q&A] [answered]
The official toturial shows the way to using an imported markdown file ([importing-markdown](https://docusaurus.io/docs/markdown-features/react#importing-markdown))
1. create a external `a.mdx` file.
2. import the `a.mdx` file in `b.mdx` by adding a line  `import FileA from './a.mdx'; `.
3. use `<FileA/>` in 'b.mdx'.

```mdx
# B.mdx
this is b.mdx

import FileA from './a.mdx';

<FileA/>

```

I think it is a littile bit complicated to import a markdown file.If I want to import athother markdown file, I just need to add a line in somewhere I want, like `{% include "a.md" %}`.That's how gitbook do in their framework.

Since we could using `Admonitions` in a common file (.md) ,could we also using some tags to import the other common MD files instead of mdx file into a common MD file? Sometimes I don't want a '.md' file to change into a '.mdx' file.

What I imagine the way to import a common markdown file like:
```markdown
# B.md
this is b.md

{ import "./a.md" }

```

Some different ways to import the markdown file:
- importing file by relative path 
- importing file by `docId`

And I am confusing about that, if the imported file is a common markdown,could we solve the rendering problem ([issue3915](https://github.com/facebook/docusaurus/issues/3915)) of importing mdx file?

Maybe a Remark/Rehype plugin can be help , but I didin't try to make it possible.Because it is a little hard for me to work on it.

**Comment by slorber** [comment, marked as answer]
> I think it is a littile bit complicated to import a markdown file.If I want to import athother markdown file, I just need to add a line in somewhere I want, like {% include "a.md" %}.That's how gitbook do in their framework.

This is not exactly the same. By importing one mdx doc into another you have 2 React components in the end. 

When using such include syntax, other tools will likely "inline" the imported markdown before compiling that single markdown document. In our case this would lead to a single final React component for the doc.

---

> Since we could using Admonitions in a common file (.md) ,could we also using some tags to import the other common MD files instead of mdx file into a common MD file? Sometimes I don't want a '.md' file to change into a '.mdx' file.

We process all md/mdx docs with MDX atm and you can keep the .md extension if you want.

---

> And I am confusing about that, if the imported file is a common markdown,could we solve the rendering problem (https://github.com/facebook/docusaurus/issues/3915) of importing mdx file?

The extension of the file does not matter: we always use MDX atm. It is the way things are imported (2 react components vs 1 react component) that makes the difference.

---

> Maybe a Remark/Rehype plugin can be help , but I didin't try to make it possible.Because it is a little hard for me to work on it.

This is definitively a good way to solve the problem. This is hard for us too, but someone has to put the hours to solve it ;)

Luckily some people seem to have worked on it already, and you'll find some initial code to test here:

https://github.com/dotansimha/remark-import-partial

https://gist.github.com/jknoxville/8ed4cb8bcef348362a99dd70bb8006ae

**Reply by AnteeHub** [reply]
Thanks for your reply. And I realize that I can use '.mdx' and '.md' at the same time.Docusaurus will processs all both docs in the same way.That's somewhere I was confused before,because the toturial doesn't  mention it in the `importing-markdown` chapter.

**Reply by pelikhan** [reply]
Is the `remark-import-partial` compatible with generating the side TOC?

**Reply by slorber** [reply]
It should work with the `beforeDefaultRemarkPlugins` option normally, but didn't try.

Note: this plugin will likely only work with MDX 1. With MDX 2 (upcoming in Docusaurus v3), another remark impl is probably going to be needed

**Reply by pelikhan** [reply]
It turns out it doesn't because it stuff the root node as a children of paragraph, which defeats the toc.
